src/HamiltonianBuilder/DOCI.cpp

```diff
@@ -53,42 +53,80 @@ Eigen::MatrixXd DOCI::constructHamiltonian(const HamiltonianParameters& hamilton
     Eigen::VectorXd diagonal = calculateDiagonal(hamiltonian_parameters);
     Eigen::MatrixXd result_matrix = Eigen::MatrixXd::Zero(dim, dim);
+    size_t N = this->fock_space.get_N();
+
     // Create the first spin string. Since in DOCI, alpha == beta, we can just treat them as one
     // And multiply all contributions by 2
     ONV onv = this->fock_space.get_ONV(0);  // spin string with address 0
 
-    for (size_t I = 0; I < dim; I++) {  // I loops over all the addresses of the spin strings
-        // Diagonal contribution
-        result_matrix(I, I) += diagonal(I);
-
-        // Off-diagonal contribution
-        for (size_t e1 = 0; e1 < this->fock_space.get_N(); e1++) {  // e1 (electron 1) loops over the (number of) electrons
-            size_t p = onv.get_occupied_index(e1);  // retrieve the index of the orbital the electron occupies
-            for (size_t q = 0; q < p; q++) {  // q loops over SOs
-                if (!onv.isOccupied(q)) {  // if q not in I
-                    onv.annihilate(p);
-                    onv.create(q);
-
-                    size_t J = this->fock_space.getAddress(onv);  // J is the address of a string that couples to I
+    for (size_t I = 0; I < dim; I++) {  // I loops over all the addresses of the onv
 
-                    // The loops are p->K and q<p. So, we should normally multiply by a factor 2 (since the summand is symmetric)
-                    // However, we are setting both of the symmetric indices of Hamiltonian, so no factor 2 is required
-                    result_matrix(I, J) += hamiltonian_parameters.get_g()(p, q, p, q);
-                    result_matrix(J, I) += hamiltonian_parameters.get_g()(p, q, p, q);
-
-                    onv.annihilate(q);  // reset the spin string after previous creation
-                    onv.create(p);  // reset the spin string after previous annihilation
+        result_matrix(I, I) += diagonal(I);
 
-                }  // q < p loop
+        for (size_t e1 = 0; e1 < N; e1++) {  // e1 (electron 1) loops over the (number of) electrons
+            size_t p = onv.get_occupied_index(e1);  // retrieve the index of a given electron
+
+            // remove the weight from the initial address I, because we annihilate
+            size_t address = I - this->fock_space.get_vertex_weights(p, e1 + 1);
+            // The e2 iteration counts the amount of encountered electrons for the creation operator
+            // We only consider greater addresses than the initial one (because of symmetry)
+            // Hence we are only required to start counting from the annihilated electron (e1)
+            size_t e2 = e1;
+
+            size_t q = p + 1;
+
+            // Test whether next orbital is occupied, until we reach unoccupied orbital
+            while (e2 < N - 1 && q == onv.get_occupied_index(e2 + 1)) {
+                // Shift the address for the electrons encountered after the annihilation but before the creation
```

Comment:
Update these lines of documentation: add capitals, make sure you form correct sentences. This feels like it's over-documented: maybe write a separate `FockSpace` function to do this step?

Reply:
Good idea.